update test cases for authenticate-dashboard-user

joel-simpson opened the pull request:
Replace 'user' with "Alice", "Bert" and "Charlie" to map to test users accounts set up (and improve readability).
Added 'given' steps to better show user set up - e.g. with specific entitlements, and to clarify the difference between and IdP user account, and an Archivematica user account. 
Add step for the information release page with IdP.

joel-simpson commented:
Just fyi - I have tested all 4 scenarios. Three pass. 
The third scenario I can't seem to recreate at the moment... I log in as Alice as per normal, then to simulate having an IdP session, but not having an Archivematica session, I delete the Archivematica.ac.uk cookies; and try another link in the dashboard. I expect to get redirected but it allows me to just carry on in Archivematica. 
Given we are going to run our own IdP initially for MVP, I'm not sure this scenario has any relevance. I'm tempted to comment it out completely for now and address it in a future iteration.  But would be interested in your thoughts @lower29 and @helenst

lower29 commented:
I agree that updating the Gherkin to use actual example users' names is a good idea, and makes it clearer to read.

I wonder however if we could make the Gherkin more succinct by using [Scenario Outlines](http://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines) to reduce the amount of repetition in the scenario definitions. For example:
```
Scenario: Preservation-User logs in with correct credentials
  Given <user> has an existing account with her identity provider 
    And <user> has the entitlement <entitlement>
    And <user> has an existing Archivematica user account 
    And <user> is not already logged in to Archivematica
  
  When <user> enters (or clicks) a URL for the Archivematica dashboard
    And Archivematica determines <user> is not already logged in
    And Archivematica redirects <user> to the Identity Provider
    And the Identity Provider determines the <user> does not have an existing authenticated session
  
  Then the Identity Provider presents a login page
    And <user> enters their user name <username> and password <password>
    And the Identity Provider authenticates <user>
    And the Identity Provider presents an Information Release consent page
    And <user> selects the option "Ask me again at next login" and clicks "accept"
    And the Identity Provider redirects <user> to Archivematica
    And Archivematica validates the response from the Identity Provider
    And <user> is logged in with <role> privileges
    And <user> is presented with the default transfer page 
    
  Examples:
    | user  | username | password | entitlement        | role       |
    | Alice | aa       | aa12345  | preservation-user  | default    |
    | Bert  | bb       | bb12345  | preservation-admin | admin      |
```

This way we can actually use the same scenario for both Alice and Bert. Charlie's scenario can be similarly parameterised.

What do you think @joel-simpson?

helenst commented:
Hi @joel-simpson - I like this, naming the users and being more explicit about the steps makes things a lot clearer.

For that third scenario, by deleting the cookies you are removing the Archivematica session but not the IdP session (which is what we want) - when you request an Archivematica link, it goes to the IdP and comes straight back again with the authentication and creates a new Archivematica session. Perhaps it doesn't give you the IdP release consent page because you're still in the same IdP session (i.e. when it says next login, it means IdP login) and going back to a service that you already gave consent for in that session? If you went to the Storage service for the first time in that session, you would see the consent page because it's a different service. (but perhaps @lower29 knows more about this aspect of IdP behaviour).

If I'm right about that, then we could either drop the assumption that it's going to show the consent page, or test the 'already logged in' scenario as part of cross-login between AM and SS.

@lower29 I like the idea of compacting things. Certainly for the admin and non admin user it makes sense, perhaps less so for the others as the flows will vary.

joel-simpson commented:
@lower29 - definitely agree with using the scenario outline. In fact I attempted that last week initially, but I think I conflated too many variables and found it was not reading any better so gave up and just cleaned up the existing scenarios. But I think the way you have done it works really well. I will update and post changes to this PR. 

@helenst - thanks for the thoughts on scenario 3. I will retest and check, I may have just missed the redirect.

joel-simpson commented:
Hi @lower29 and @helenst  
I have incorporated Richards changes for the Scenario Outline.
I have also retested the final scenario and Helen was correct... it was redirecting to IdP (just very quickly so you had to watch fast to see it work). I have added some comments to that scenario... the behaviour will change depending on which options you select on the information consent page. 
I can confirm all of these scenarios pass as expected. 
I would like to merge this pull request now, and then use this Gherkin as the template for the Storage Service gherkin.

lower29 commented:
Looks good to me :+1
